charms/tensorboards-web-app/src/charm.py

```diff
@@ -1,158 +1,252 @@
+            "description": "Pebble config layer for jupyter-web-app",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "Entrypoint of jupyter-web-app image",
+                    "command": exec_command,
+                    "startup": "enabled",
+                    "environment": self._env_vars,
+                    "on-check-failure": {"up": "restart"},
+                }
+            },
+            "checks": {
+                "up": {
+                    "override": "replace",
+                    "period": "30s",
+                    "http": {"url": f"http://localhost:{self._http_port}"},
                 },
-                "containers": [
-                    {
-                        "name": "tensorboards-web-app",
-                        "imageDetails": image_details,
-                        "envConfig": {
-                            "USERID_HEADER": "kubeflow-userid",
-                            "USERID_PREFIX": "",
-                            "APP_PREFIX": "/tensorboards",
-                            "APP_SECURE_COOKIES": str(config["secure-cookies"]),
-                            "BACKEND_MODE": config["backend-mode"],
-                        },
-                        "ports": [{"name": "http", "containerPort": config["port"]}],
-                    }
-                ],
             },
-        )
+        }
 
-        self.model.unit.status = ActiveStatus()
+        return Layer(layer_config)
+
+    def _on_install(self, _) -> None:
+        """Perform installation only actions."""
+        try:
+            # Deploy K8S resources to speed up deployment
+            self._deploy_k8s_resources()
+        except ErrorWithStatus as err:
+            self._log_and_set_status(err.status)
+        return
+
+    def _on_remove(self, _) -> None:
+        """Remove all resources."""
+        self._log_and_set_status(MaintenanceStatus("Removing K8S resources"))
+        k8s_resources_manifests = self.k8s_resource_handler.render_manifests()
+        try:
+            delete_many(self.k8s_resource_handler.lightkube_client, k8s_resources_manifests)
+        except ApiError as err:
+            # Do not log/report when resources were not found
+            if err.status.code != 404:
+                self.logger.error(f"Removing K8S resources failed with error: {err}")
+                raise GenericCharmRuntimeError("Removing K8s resources failed") from err
+
+    def _on_event(self, event) -> None:
+        """Perform required actions for every event."""
+        try:
+            self._check_leader()
+            self._deploy_k8s_resources()
+            if self._is_container_ready():
+                update_layer(
+                    self._container_name,
+                    self._container,
+                    self._tensorboards_web_app_layer,
+                    self.logger,
+                )
```

Comment:
Good observation, I think I can just remove the `if` statement since all the functions are inside in a `try..except` statement and [update_layer()](https://github.com/canonical/charmed-kubeflow-chisme/blob/c82bd1dd21b375c6cf91f43298fdc09d98e4e07c/src/charmed_kubeflow_chisme/pebble/_update_layer.py#L22-L24) raises an error based on `can_connect()`. WDYT? @DnPlas

Reply:
Yes. update_layer() should catch and raise that. I think it is safe to remove it.